addon/services/socket.js

```diff
@@ -0,0 +1,157 @@
+    showDisconnectAtStep: 4,
+    reconnectOnDate: null,
+    isConnecting: false,
+    lostConnection: false,
+
+    showDisconnect: computed('reconnectDelayStep', 'showDisconnectAtStep', function() {
+        return get(this, 'reconnectDelayStep') >= get(this, 'showDisconnectAtStep')
+    }),
+
+    init() {
+        this._super(...arguments);
+        set(this, 'subscriptions', {});
+    },
+
+    host: computed('baseURL', 'requiresAuth', 'session.data.authenticated.access_token', function() {
+        let host = get(this, 'baseURL');
+        const token = get(this, 'session.data.authenticated.access_token');
+        const requiresAuth = get(this, 'requiresAuth');
+        if(requiresAuth && token) {
+            host += `?access_token=${token}`;
+        }
+
+        return host;
+    }),
+
+    connect() {
+        if(!get(this, 'fastboot.isFastBoot')) {
+            return new Promise((resolve, reject) => {
+                //dont attempt to connect if already connected
+                //or in the process of connecting
+                if(get(this, 'client.connected')) {
+                    return resolve(get(this, 'client'));
+                }
+
+                if(get(this, 'isConnecting')) {
+                    return reject();
+                }
+
+                const client = Stomp.over(() => {
+                    return new SockJS(get(this, 'host'));
+                });
+
+                //handling implementation of auto-reconnect in the app (to support event notifications)
+                client.reconnect_delay = 0;
+
+                //allow console log debug output to be disabled
+                if(!get(this, 'debug')) {
+                    client.debug = null;
+                }
+
+                cancel(get(this, 'reconnectTimer'));
+                setProperties(this, {
+                    reconnectOnDate: null,
+                    isConnecting: true,
+                    client
+                });
+
+                client.connect({},
+                    () => {
+                        get(this, 'eventBus').publish('socketConnected');
+                        set(this, 'lostConnection', false);
+                        setProperties(this, {
+                            lostConnection: false,
+                            isConnecting: false,
+                            reconnectDelayStep: 0
+                        });
+
+                        return resolve(client);
+                    },
+                    (error) => {
+                        const reconnectStep = get(this, 'reconnectDelayStep');
+                        const numSteps = get(this, 'reconnectDelaySteps.length');
+                        //keep increasing the reconnect delay, until the max delay is reached
+                        const reconnectDelayStep = (reconnectStep === numSteps - 1) ? reconnectStep : (reconnectStep + 1);
+                        get(this, 'eventBus').publish('socketDisconnected', error);
```

Comment:
@billdami another benefit of using the socket service to handle global events is you don't need to prefix the event name with `socket` anymore. 

```
get(this, 'eventBus').publish('socketDisconnected', error);
```
would become
```
get(this, 'socket').publish('disconnected', error);
```